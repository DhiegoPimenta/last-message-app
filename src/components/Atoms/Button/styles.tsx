import styled, { css, DefaultTheme } from 'styled-components'
import { ButtonProps } from '.'

const wrapperModifiers = {
    success: (theme: DefaultTheme) => css`
        background-color: ${theme.colors.primary};
        color: ${theme.colors.white};
    `,
    danger: (theme: DefaultTheme) => css`
        background-color: ${theme.colors.danger};
        color: ${theme.colors.white};
    `,
    info: (theme: DefaultTheme) => css`
        background-color: ${theme.colors.gray};
        color: ${theme.colors.white};
    `,
}

export const Button = styled.button<ButtonProps>`
    ${({ theme, colorStyle }) => css`
        ${wrapperModifiers[colorStyle!](theme)}
        border-radius: ${theme.border.radius};
        width: 18rem;
        height: 5rem;
        border: 0;
        cursor: pointer;

        font-weight: ${theme.font.light};
        font-size: ${theme.font.sizes.medium};
    `})}
`
