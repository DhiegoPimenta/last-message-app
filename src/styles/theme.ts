export default {
    grid: {
        container: '130rem',
        gutter: '3.2rem',
    },
    border: {
        radius: '1rem',
    },
    font: {
        family: `Poppins, Ubuntu, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif`,
        light: 300,
        normal: 400,
        bold: 600,
        extrabold: 800,
        sizes: {
            xsmall: '1.4rem',
            small: '1.8rem',
            medium: '2.4rem',
            large: '3.2rem',
        },
    },
    colors: {
        primary: '#64CBC5',
        secondary: '#0030FF',
        terciary: '#A0DFDC',
        white: '#FFFFFF',
        gray: '#CECECE',
        danger: '#FBA9B0',
        black: '#444444',
        overlay: 'rgba(0, 0, 0, 0.3)',
    },
    spacings: {
        xxsmall: '0.8rem',
        xsmall: '1.6rem',
        small: '2.4rem',
        medium: '3.2rem',
        large: '4.0rem',
        xlarge: '4.8rem',
        xxlarge: '5.6rem',
    },
    layers: {
        base: 10,
        menu: 20,
        overlay: 30,
        modal: 40,
        alwaysOnTop: 50,
    },
    transition: {
        default: '0.3s ease-in-out',
        fast: '0.1s ease-in-out',
    },
} as const
