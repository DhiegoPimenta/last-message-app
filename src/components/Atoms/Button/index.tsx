import * as S from './styles'

export type ButtonColorStyle = 'info' | 'success' | 'danger'

export interface ButtonProps
    extends React.ButtonHTMLAttributes<HTMLButtonElement> {
    children: React.ReactNode
    colorStyle?: ButtonColorStyle
    isLoading?: boolean
}

const Button = ({
    children,
    colorStyle = 'success',
    type,
    isLoading,
    ...props
}: ButtonProps) => (
    <S.Button colorStyle={colorStyle} type={type} {...props}>
        {isLoading ? 'enviando...' : children}
    </S.Button>
)

export default Button
