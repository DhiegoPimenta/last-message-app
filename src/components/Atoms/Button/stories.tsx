import { Story, Meta } from '@storybook/react/types-6-0'

import Button, { ButtonProps } from '.'

export default {
    title: 'Atoms/Button',
    component: Button,
} as Meta

export const Basic: Story<ButtonProps> = (args) => (
    <Button {...args}>Click me</Button>
)
Basic.args = {
    colorStyle: 'success',
}
