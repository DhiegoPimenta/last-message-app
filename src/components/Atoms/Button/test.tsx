import { fireEvent, screen } from '@testing-library/react'
import theme from 'styles/theme'
import { renderWithTheme } from 'utils/tests/helpers'
import Button from '.'

describe('<Button/>', () => {
    it('Should be render', () => {
        const { container } = renderWithTheme(<Button>Button</Button>)

        expect(screen.getByText(/Button/i)).toBeInTheDocument()
        expect(container.firstChild).toMatchSnapshot()
    })

    it('Should call the funtion onclick', () => {
        const mockClick = jest.fn()
        renderWithTheme(<Button onClick={mockClick}>Button</Button>)

        fireEvent.click(screen.getByText(/Button/i))

        expect(screen.getByText(/Button/i)).toHaveStyle({
            cursor: 'pointer',
        })
        expect(mockClick).toHaveBeenCalled()
    })

    it('Should be primary background color as default', () => {
        renderWithTheme(<Button>Button</Button>)

        expect(screen.getByText(/button/i)).toHaveStyle({
            'background-color': theme.colors.primary,
            color: theme.colors.white,
        })
    })
})
