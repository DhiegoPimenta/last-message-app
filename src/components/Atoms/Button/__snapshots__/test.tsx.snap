// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> Should be render 1`] = `
.c0 {
  background-color: #64CBC5;
  color: #FFFFFF;
  border-radius: 1rem;
  width: 18rem;
  height: 5rem;
  border: 0;
  cursor: pointer;
  font-weight: 300;
  font-size: 2.4rem;
}

<button
  class="c0"
>
  Button
</button>
`;
